### PR TITLE
Ensure that exiting thread invokes end-of-life behaviour.

### PR DESCRIPTION
**Pull request by ioquatix**
https://bugs.ruby-lang.org/issues/20286

### CODE REVIEW

File: thread.c [hunk 1]
```diff
@@ -662,12 +656,31 @@ thread_start_func_2(rb_thread_t *th, VALUE *stack_start)
     // Ensure that we are not joinable.
     VM_ASSERT(UNDEF_P(th->value));
 
+    int rb_fiber_scheduler_set_invoked = 0, event_thread_end_hooked = 0;
```
Comment (ko1) [inline thread.c:659]: do not use `rb_` prefix for local variables.

File: thread.c [hunk 2]
```diff
@@ -662,12 +656,31 @@ thread_start_func_2(rb_thread_t *th, VALUE *stack_start)
...
-        SAVE_ROOT_JMPBUF(th, thread_do_start(th));
+        EXEC_EVENT_HOOK(th->ec, RUBY_EVENT_THREAD_BEGIN, th->self, 0, 0, 0, Qundef);
+
+        SAVE_ROOT_JMPBUF(th, result = thread_do_start(th));
     }
-    else {
+
+    if (!rb_fiber_scheduler_set_invoked) {
+        rb_fiber_scheduler_set(Qnil);
+        rb_fiber_scheduler_set_invoked = 1;
```
[the remark this answers is not in the document]
Comment (ioquatix) [reply thread.c:672]: Ah, the order is wrong.

File: thread.c (hunk 1 above)
Comment (ioquatix) [reply thread.c:659]: I don't really mind what we call it.

File: thread.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ioquatix) [reply thread.c:671]: Fixed.

**Pull request merged**